### PR TITLE
[Improve][Connector-v2] Optimize the way of databases and tables are checked for existence

### PR DESCRIPTION
**Pull request by dailai**
### Purpose of this pull request

1. Changed the way about all jdbc catalogs checked for the db or table exists. This is because the original method of checking whether a table exists is to simply list all the tables and then check whether all the table names contain the given table name, which can be a performance problem.

2. In addition to the above issues, in mysql, it is not possible to create tables with the same table name in different case in default, which can lead to errors like the following
image

### Check list

* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/seatunnel/tree/dev/docs
* [ ] If you are contributing the connector code, please check that the following files are updated:
  1. Update change log that in connector document. For more details you can refer to [connector-v2](https://github.com/apache/seatunnel/tree/dev/docs/en/connector-v2)
  2. Update [plugin-mapping.properties](https://github.com/apache/seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  3. Update the pom file of [seatunnel-dist](https://github.com/apache/seatunnel/blob/dev/seatunnel-dist/pom.xml)
* [ ] Update the [`release-note`](https://github.com/apache/seatunnel/blob/dev/release-note.md).

### CODE REVIEW

**Comment by dailai** [comment]
1. Mysql
mysql_catalog

2. Oracle
oracle_catalog

3. Postgresql
postgre_catalog

4. Sqlserver
sqlserver_catalog

In addition, org.apache.seatunnel.connectors.seatunnel.jdbc.AbstractJdbcIT#testCatalog will verify related e2e case.

**Comment by hailin0** [comment]
Thanks，can you change all catalogs?

https://github.com/apache/seatunnel/tree/dev/seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog

**Comment by dailai** [comment]
> Thanks，can you change all catalogs?
> 
> https://github.com/apache/seatunnel/tree/dev/seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog

done

**Comment by dailai** [comment]
> Thanks，can you change all catalogs?
> 
> https://github.com/apache/seatunnel/tree/dev/seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog

done

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/redshift/RedshiftCatalog.java
Comment (hailin0) [inline seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/redshift/RedshiftCatalog.java]: remove ';' from `getListDatabaseSql()`

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/redshift/RedshiftCatalog.java
Comment (hailin0) [inline seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/redshift/RedshiftCatalog.java:84]: remove ';' from `getListTableSql()`

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/dm/DamengCatalog.java
Comment (hailin0) [inline seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/dm/DamengCatalog.java]:
```suggestion
        return String.format(getListDatabaseSql() + " where name = '%s'", databaseName);
```

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/dm/DamengCatalog.java
Comment (hailin0) [inline seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/dm/DamengCatalog.java]:
```suggestion
                        + " where OWNER = '%s' and TABLE_NAME = '%s'",
```

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/iris/IrisCatalog.java
Comment (hailin0) [inline seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/iris/IrisCatalog.java]:
```suggestion
                getListTableSql(tablePath.getSchemaName()) + " and TABLE_NAME = '%s'",
```

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/oracle/OracleCatalog.java
Comment (hailin0) [inline seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/oracle/OracleCatalog.java]:
```suggestion
                + "  and  OWNER = '"
```

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/psql/PostgresCatalog.java
Comment (hailin0) [inline seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/psql/PostgresCatalog.java:114]:
```suggestion
                        + " where table_schema = '%s' and table_name = '%s'",
```

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/redshift/RedshiftCatalog.java
Comment (hailin0) [inline seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/redshift/RedshiftCatalog.java]:
```suggestion
        return String.format(getListDatabaseSql() + " where datname = '%s'", databaseName);
```

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/redshift/RedshiftCatalog.java
Comment (hailin0) [inline seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/redshift/RedshiftCatalog.java]:
```suggestion
                        + " where table_schema = '%s' and table_name = '%s'",
```

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/saphana/SapHanaCatalog.java
Comment (hailin0) [inline seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/saphana/SapHanaCatalog.java]:
```suggestion
        return String.format(getListDatabaseSql() + " where SCHEMA_NAME = '%s'", databaseName);
```

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/saphana/SapHanaCatalog.java
Comment (hailin0) [inline seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/saphana/SapHanaCatalog.java]:
```suggestion
                getListTableSql(tablePath.getDatabaseName()) + " and TABLE_NAME = '%s'",
```

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/sqlserver/SqlServerCatalog.java
Comment (hailin0) [inline seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/sqlserver/SqlServerCatalog.java]:
```suggestion
        return String.format(getListDatabaseSql() + "  where name = '%s'", databaseName);
```

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/sqlserver/SqlServerCatalog.java
Comment (hailin0) [inline seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/sqlserver/SqlServerCatalog.java]:
```suggestion
                        + "  and  TABLE_SCHEMA = '%s' and TABLE_NAME = '%s'",
```

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/xugu/XuguCatalog.java
Comment (hailin0) [inline seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/xugu/XuguCatalog.java]:
```suggestion
        return String.format(getListDatabaseSql() + "  where DB_NAME = '%s'", databaseName);
```

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/xugu/XuguCatalog.java
Comment (hailin0) [inline seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/xugu/XuguCatalog.java]:
```suggestion
                        + "  where user_name = '%s' and table_name = '%s'",
```

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/redshift/RedshiftCatalog.java
Comment (dailai) [reply seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/redshift/RedshiftCatalog.java]: done

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/AbstractJdbcCatalog.java
Comment (Hisoka-X) [inline seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/AbstractJdbcCatalog.java]:
```suggestion
        if (SYS_DATABASES.contains(databaseName)) {
```

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/AbstractJdbcCatalog.java
Comment (Hisoka-X) [inline seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/AbstractJdbcCatalog.java]:
```suggestion
    protected boolean querySQLResultExists(String dbUrl, String sql) {
```

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/AbstractJdbcCatalog.java
Comment (Hisoka-X) [inline seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/AbstractJdbcCatalog.java]:
```suggestion
        String databaseName = tablePath.getDatabaseName();
        if (EXCLUDED_SCHEMAS.contains(tablePath.getSchemaName())) {
            return false;
        }
        return queryExists(
                this.getUrlFromDatabaseName(databaseName), getTableWithConditionSql(tablePath));
```
how about remove check database exist? It won't affect the result.

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/AbstractJdbcCatalog.java
Comment (Hisoka-X) [inline seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/AbstractJdbcCatalog.java]: Do we need to keep the rollback strategy? If Catalog does not support the `getDatabaseWithConditionSql` method, we can continue to use the old check way.

Comment (dailai) [reply seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/AbstractJdbcCatalog.java]: > Do we need to keep the rollback strategy? If Catalog does not support the `getDatabaseWithConditionSql` method, we can continue to use the old check way.

I think it's no need to use this strategy because  all jdbc catalogs are supported this way.

Comment (Hisoka-X) [reply seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/AbstractJdbcCatalog.java]: But if the user adds a custom Catalog, it needs to be re-adapted. This is my concern.

Comment (dailai) [reply seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/AbstractJdbcCatalog.java]: > But if the user adds a custom Catalog, it needs to be re-adapted. This is my concern.

I think the old way is a wrong way to use, should be abandoned, if the user needs to customize the catalog should be a need to consider this issue.

Comment (Hisoka-X) [reply seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/AbstractJdbcCatalog.java]: But in fact, the old way only has performance issues. If the user does not have a lot of tables, there is no problem. This PR solves the performance problem but should not invalidate the functions of the custom Catalog (It can not worked after merge this PR if do not change the code).

Comment (dailai) [reply seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/AbstractJdbcCatalog.java]: > But in fact, the old way only has performance issues. If the user does not have a lot of tables, there is no problem. This PR solves the performance problem but should not invalidate the functions of the custom Catalog (It can not worked after merge this PR if do not change the code).

In fact, even the original method is also executed the SQL first to get all the table names, and if the SQL  is not supported, the original method is also not supported. I just added a where condition to the original sql, it doesn't seem to make any difference

Comment (Hisoka-X) [reply seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/AbstractJdbcCatalog.java]: The old Catalog doesn't implement `getDatabaseWithConditionSql` method. So when invoke this method, the `UnsupportedOperationException` would be throwed.

The difference is that the user need or no need to modify the code

Comment (dailai) [reply seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/AbstractJdbcCatalog.java]: > The old Catalog doesn't implement `getDatabaseWithConditionSql` method. So when invoke this method, the `UnsupportedOperationException` would be throwed.
> 
> The difference is that the user need or no need to modify the code

Okay, you convinced me.

**Review by Hisoka-X — APPROVED** [review]
LGTM

**Review by hailin0 — APPROVED** [review]

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/oceanbase/OceanBaseOracleCatalog.java
Comment (Hisoka-X) [inline seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/oceanbase/OceanBaseOracleCatalog.java:58]: Hi @dailai  ,Why oracle support `getDatabaseWithConditionSql` but OceanBaseOracle can not support `getDatabaseWithConditionSql`?

Comment (dailai) [reply seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/oceanbase/OceanBaseOracleCatalog.java:58]: > Hi @dailai ,Why oracle support `getDatabaseWithConditionSql` but OceanBaseOracle can not support `getDatabaseWithConditionSql`?

Beacase this method named getListDatabaseSql is not supported.

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/AbstractJdbcCatalog.java
Comment (Hisoka-X) [inline seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog/AbstractJdbcCatalog.java:583]: Why we should catch all exception we query failed? This is a mistake.

**Pull request merged**
